src/Symfony/Component/Mailer/Transport/AbstractHttpTransport.php

```diff
@@ -62,13 +62,13 @@ public function setPort(?int $port): static
         return $this;
     }
 
-    abstract protected function doSendHttp(SentMessage $message): ResponseInterface;
+    abstract protected function doSendHttp(SentMessage $message): ?ResponseInterface;
 
     protected function doSend(SentMessage $message): void
     {
         try {
             $response = $this->doSendHttp($message);
-            $message->appendDebug($response->getInfo('debug') ?? '');
+            $message->appendDebug($response?->getInfo('debug') ?? '');
```

Comment:
here is the BC break in practice; if you don't do this change: :boom: - any child class can be affected in the same way

Reply:
Yes ; but only if Symfony implementations start to really returning `null` instead of a Response, which is not the case currently.
And the day someone start to use an implementation which returns null (by installing this library) they will have to support null, but this won't be considered as a BC break since they install a new lib.

Reply:
Once the return type of AbstractApiTransport say it is `ResponseInterface|null`, existing implementation would not see a BC break if they change a returned value while respecting this contract, as this would not change the contract.
The BC break happens when you change the contract (the fact that the runtime failure does not happen immediately in all case is a different matter).

Reply:
Should I close this then or is something else possible ?

Reply:
To me, this should be closed. Not doing an HTTP request in the method `doSendHttp` does not make sense to me.

Reply:
> To me, this should be closed. Not doing an HTTP request in the method `doSendHttp` does not make sense to me.

First, you can do an HttpRequest with something else than an HttpClientInterface which doesn't return an ResponseInterface (and there is no easy way to re-create something implementing the response interface)

Secondly, my issue was more about the `AbstractApiTransport`.
This abstract class provide useful things like
- an email param
- a getRecipients method

And again you can do API calls with custom SDK, which does not implements HttpClientInterface and therefore doesn't return a ResponseInterface.

Reply:
Well, `AbstractHttpTransport` depends on `symfony/http-client-contracts`. That's why it expects the ResponseInterface of that package. If you want to build things on your own, do it without the base class IMO.

> This abstract class provide useful things like
> 
>     * an email param
> 
>     * a getRecipients method

If you want this Email object without the parent `AbstractHttpTransport`, you could reuse `\Symfony\Component\Mime\MessageConverter::toEmail` directly (which is how this `$email` is obtained).

For `getRecipients`, this is just a helper filtering the Envelope recipients to remove the recipients coming from Cc and Bcc (as APIs will take those separately generally).